### PR TITLE
Set http.route in spring-autoconfigure webmvc instrumentation

### PR DESCRIPTION
**Pull request by mateuszrzeszutek**
Resolves https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/3479

I think this feature should be integrated into the `spring-webmvc` library instrumentation; it'd require bumping the minimum version to 5.3 though. I think we should do that, having the route is super useful, and the webmvc library instrumentation is probably only used in autoconfigure anyway.

### CODE REVIEW

**Comment by trask** [comment]
> I think this feature should be integrated into the `spring-webmvc` library instrumentation; it'd require bumping the minimum version to 5.3 though. I think we should do that, having the route is super useful, and the webmvc library instrumentation is probably only used in autoconfigure anyway.

👍 to bumping the min version for spring-webmvc library instrumentation

do you want to merge this PR first, or make that change here?

**Comment by mateuszrzeszutek** [comment]
I could do this in this PR, I don't mind either way

File: instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpCommonAttributesGetter.java [hunk 1]
```diff
@@ -58,9 +58,21 @@ default Long requestContentLengthUncompressed(REQUEST request, @Nullable RESPONS
...
+  /**
+   * Extracts the {@code http.status_code} span attribute.
+   *
+   * <p>This is called from {@link Instrumenter#end(Context, Object, Object, Throwable)}, only when
+   * {@code response} is non-{@code null}.
+   */
+  @Nullable
+  default Integer statusCode(REQUEST request, RESPONSE response, @Nullable Throwable error) {
+    return statusCode(request, response);
+  }
```
Comment (mateuszrzeszutek) [inline instrumentation-api-semconv/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/http/HttpCommonAttributesGetter.java:74]: I wanted the Spring/Servlet status code to be properly reflected in the span status, so I added this method. I'll deprecate the previous one in the next PR.

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java [hunk 2]
```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright The OpenTelemetry Authors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package io.opentelemetry.instrumentation.spring.webmvc.v5_3;
```
Comment (mateuszrzeszutek) [inline instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:6]: I also changed the package name, in accordance with #932

Comment (trask) [reply instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:6]: I had to think through this (https://github.com/open-telemetry/opentelemetry-java-instrumentation/issues/932#issuecomment-1206678184), but I agree

File: instrumentation/spring/spring-webmvc-5.3/library/src/test/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/WebMvcHttpServerTest.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+
+import io.opentelemetry.instrumentation.testing.junit.InstrumentationExtension;
+import io.opentelemetry.instrumentation.testing.junit.http.AbstractHttpServerTest;
+import io.opentelemetry.instrumentation.testing.junit.http.HttpServerInstrumentationExtension;
+import io.opentelemetry.instrumentation.testing.junit.http.HttpServerTestOptions;
+import io.opentelemetry.instrumentation.testing.junit.http.ServerEndpoint;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.springframework.context.ConfigurableApplicationContext;
+
+class WebMvcHttpServerTest extends AbstractHttpServerTest<ConfigurableApplicationContext> {
```
Comment (trask) [inline instrumentation/spring/spring-webmvc-5.3/library/src/test/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/WebMvcHttpServerTest.java:16]: nice!

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+import org.springframework.web.context.support.WebApplicationContextUtils;
+import org.springframework.web.servlet.DispatcherServlet;
+import org.springframework.web.servlet.HandlerExecutionChain;
+import org.springframework.web.servlet.HandlerMapping;
+import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping;
+import org.springframework.web.util.ServletRequestPathUtils;
+
+final class HttpRouteSupport {
+
+  private final AtomicBoolean contextRefreshTriggerred = new AtomicBoolean();
```
Comment (trask) [inline instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:33]:
```suggestion
  private final AtomicBoolean contextRefreshTriggered = new AtomicBoolean();
```

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+      ((ConfigurableWebApplicationContext) context)
+          .addApplicationListener(new WebContextRefreshListener());
+    }
+  }
+
+  // we can't retrieve the handler mappings from the DispatcherServlet in the onRefresh listener,
+  // because it loads them just after the application context refreshed event is processed
+  // to work around this, we're setting a boolean flag that'll cause this filter to load the
+  // mappings the next time it attempts to set the http.route
+  final class WebContextRefreshListener implements ApplicationListener<ContextRefreshedEvent> {
```
Comment (trask) [inline instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:58]: just curious if this gets hit in the integration tests (I'm ok if not)

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+import org.springframework.web.servlet.DispatcherServlet;
+import org.springframework.web.servlet.HandlerExecutionChain;
+import org.springframework.web.servlet.HandlerMapping;
+import org.springframework.web.servlet.mvc.method.annotation.RequestMappingHandlerMapping;
+import org.springframework.web.util.ServletRequestPathUtils;
+
+final class HttpRouteSupport {
+
+  private final AtomicBoolean contextRefreshTriggerred = new AtomicBoolean();
+  @Nullable private WeakReference<DispatcherServlet> dispatcherServlet;
```
Comment (trask) [inline instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:34]: can you add brief comment why WeakReference is needed?

**Review by trask — APPROVED** [review]

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java (hunk 4 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:33]: The intricacies of the English language... 🙈 
(On a side note, I wonder why IntelliJ did not highlight this)

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java (hunk 5 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:58]: It does; it has to be hit if the `http.route` is to be present in generated spans.

File: instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java (hunk 6 above)
Comment (mateuszrzeszutek) [reply instrumentation/spring/spring-webmvc-5.3/library/src/main/java/io/opentelemetry/instrumentation/spring/webmvc/v5_3/HttpRouteSupport.java:34]: I think it wasn't needed after all; removed.

**Pull request merged**